File: android/lib/src/main/kotlin/io/grin/lib/GrinPlayer.kt
package io.grin.lib

class GrinPlayer(
    private val schedulerFactory: (Long) -> TickScheduler = { tickMicros ->
        AndroidTickScheduler(tickMicros)
    },
    private val ruleEngine: RuleEngine = RuleEngine(),
    private val opcodeRegistry: OpcodeRegistry = OpcodeRegistry,
    private val onFrameRendered: (() -> Unit)? = null
) {
    private var image: GrinImage? = null
    private var scheduler: TickScheduler? = null
    private var state: PlaybackState? = null
    private var displayBuffer: DisplayBuffer? = null
    private var controlBytes: IntArray = IntArray(0)

    fun load(file: GrinFile) {
        val header = file.header
        val width = safeInt(header.width, "Width")
        val height = safeInt(header.height, "Height")

        image = GrinImage(header, file.pixels.toMutableList(), file.rules)
        displayBuffer = DisplayBuffer(width, height)
        state = PlaybackState(displayBuffer!!)
        controlBytes = IntArray(file.pixels.size) { index ->
            file.pixels[index].c and GrinControlByte.RESERVED_MASK.inv() and 0xFF
        }

        scheduler?.stop()
        scheduler = schedulerFactory(header.tickMicros).also { sched ->
            sched.setTickCallback { tick -> onTick(tick) }
        }
    }

    fun play() {
        val currentState = state ?: throw IllegalStateException("GrinPlayer.load() must be called before play()")
        val currentScheduler = scheduler ?: throw IllegalStateException("Scheduler not initialized")
        if (currentState.isPlaying) {
            return
        }
        currentState.isPlaying = true
        currentScheduler.start()
    }

    fun pause() {
        val currentState = state ?: return
        val currentScheduler = scheduler ?: return
        currentState.isPlaying = false
        currentScheduler.stop()
    }

    fun stop() {
        state?.reset()
        pause()
    }

    fun seek(tick: Long) {
        require(tick >= 0L) { "Tick must be non-negative" }
        state?.currentTick = tick
        renderFrame(tick)
    }

    fun getCurrentFrame(): DisplayBuffer {
        return displayBuffer ?: throw IllegalStateException("GrinPlayer.load() must be called before getCurrentFrame()")
    }

    fun isPlaying(): Boolean {
        return state?.isPlaying ?: false
    }

    private fun onTick(tick: Long) {
        state?.currentTick = tick
        renderFrame(tick)
    }

    private fun renderFrame(tick: Long) {
        val currentImage = image ?: return
        val buffer = displayBuffer ?: return
        val output = buffer.rgbaData
        val opcodeSetId = currentImage.header.opcodeSetId
        val activeRules = ruleEngine.evaluateRules(currentImage, tick)

        currentImage.pixels.forEachIndexed { index, source ->
            val control = controlBytes[index]
            val outputIndex = index * 4

            if ((control and GrinControlByte.LOCK_MASK) != 0) {
                output[outputIndex] = source.r.toByte()
                output[outputIndex + 1] = source.g.toByte()
                output[outputIndex + 2] = source.b.toByte()
                output[outputIndex + 3] = source.a.toByte()
                return@forEachIndexed
            }

            val working = GrinPixel(source.r, source.g, source.b, source.a, control)
            val groupId = control and GrinControlByte.GROUP_ID_MASK

            for (activeRule in activeRules) {
                if (activeRule.rule.targetsGroup(groupId)) {
                    val opcode = opcodeRegistry.getOpcode(opcodeSetId, activeRule.rule.opcode)
                    opcode.apply(working, tick, activeRule.rule.timing)
                }
            }

            controlBytes[index] = working.c and GrinControlByte.RESERVED_MASK.inv() and 0xFF
            output[outputIndex] = working.r.toByte()
            output[outputIndex + 1] = working.g.toByte()
            output[outputIndex + 2] = working.b.toByte()
            output[outputIndex + 3] = working.a.toByte()
        }
        onFrameRendered?.invoke()
    }
}

private fun safeInt(value: Long, label: String): Int {
    require(value <= Int.MAX_VALUE.toLong()) { "$label exceeds Int range" }
    require(value >= 0L) { "$label must be non-negative" }
    return value.toInt()
}

File: web/lib/grin-player.ts
import { CONTROL_BYTE_MASKS } from "./format";
import { GrinFile } from "./grin-file";
import { GrinImage } from "./grin-image";
import { GrinPixel } from "./grin-pixel";
import { DisplayBuffer } from "./display-buffer";
import { OpcodeRegistry } from "./opcode-registry";
import { PlaybackState } from "./playback-state";
import { RuleEngine } from "./rule-engine";
import { RAFTickScheduler, TickScheduler } from "./tick-scheduler";

export class GrinPlayer {
  private image: GrinImage | null = null;
  private scheduler: TickScheduler | null = null;
  private readonly schedulerFactory: (tickMicros: number) => TickScheduler;
  private readonly ruleEngine: RuleEngine;
  private readonly registry: OpcodeRegistry;
  private controlBytes = new Uint8Array(0);
  private state: PlaybackState | null = null;
  private displayBuffer: DisplayBuffer | null = null;
  private readonly onFrameRendered: (() => void) | null;

  constructor(
    schedulerFactory: (tickMicros: number) => TickScheduler = (tickMicros) =>
      new RAFTickScheduler(tickMicros),
    ruleEngine: RuleEngine = new RuleEngine(),
    registry: OpcodeRegistry = OpcodeRegistry.getInstance(),
    onFrameRendered: (() => void) | null = null
  ) {
    this.schedulerFactory = schedulerFactory;
    this.ruleEngine = ruleEngine;
    this.registry = registry;
    this.onFrameRendered = onFrameRendered;
  }

  load(file: GrinFile): void {
    this.image = new GrinImage(file.header, file.pixels, file.rules);
    this.displayBuffer = new DisplayBuffer(this.image.header.width, this.image.header.height);
    this.state = new PlaybackState(this.displayBuffer);
    this.controlBytes = new Uint8Array(this.image.pixels.length);
    this.image.pixels.forEach((pixel, index) => {
      this.controlBytes[index] = pixel.c & ~CONTROL_BYTE_MASKS.RESERVED;
    });

    if (this.scheduler) {
      this.scheduler.stop();
    }
    this.scheduler = this.schedulerFactory(this.image.header.tickMicros);
    this.scheduler.setTickCallback((tick) => this.onTick(tick));
  }

  play(): void {
    if (!this.state || !this.scheduler) {
      throw new Error("GrinPlayer.load() must be called before play()");
    }
    if (this.state.isPlaying) {
      return;
    }
    this.state.isPlaying = true;
    this.scheduler.start();
  }

  pause(): void {
    if (!this.state || !this.scheduler) {
      return;
    }
    this.state.isPlaying = false;
    this.scheduler.stop();
  }

  stop(): void {
    if (!this.state) {
      return;
    }
    this.pause();
    this.state.reset();
  }

  seek(tick: number): void {
    if (!this.state || !this.image) {
      throw new Error("GrinPlayer.load() must be called before seek()");
    }
    if (!Number.isInteger(tick) || tick < 0) {
      throw new RangeError("Tick must be a non-negative integer");
    }
    this.state.currentTick = tick;
    this.renderFrame(tick);
  }

  getCurrentFrame(): DisplayBuffer {
    if (!this.displayBuffer) {
      throw new Error("GrinPlayer.load() must be called before getCurrentFrame()");
    }
    return this.displayBuffer;
  }

  isPlaying(): boolean {
    return this.state?.isPlaying ?? false;
  }

  getCurrentTick(): number {
    return this.state?.currentTick ?? 0;
  }

  private onTick(tick: number): void {
    if (!this.state) {
      return;
    }
    this.state.currentTick = tick;
    this.renderFrame(tick);
  }

  private renderFrame(tick: number): void {
    if (!this.image || !this.displayBuffer) {
      return;
    }
    const activeRules = this.ruleEngine.evaluateRules(this.image, tick);
    const output = this.displayBuffer.rgbaData;
    const opcodeSetId = this.image.header.opcodeSetId;

    for (let i = 0; i < this.image.pixels.length; i += 1) {
      const source = this.image.pixels[i];
      const control = this.controlBytes[i] ?? source.c;
      const outputIndex = i * 4;

      if ((control & CONTROL_BYTE_MASKS.LOCK) !== 0) {
        output[outputIndex] = source.r;
        output[outputIndex + 1] = source.g;
        output[outputIndex + 2] = source.b;
        output[outputIndex + 3] = source.a;
        continue;
      }

      const working = new GrinPixel(source.r, source.g, source.b, source.a, control);
      const groupId = control & CONTROL_BYTE_MASKS.GROUP_ID;

      for (const activeRule of activeRules) {
        if (activeRule.rule.targetsGroup(groupId)) {
          const opcode = this.registry.getOpcode(opcodeSetId, activeRule.rule.opcode);
          opcode.apply(working, tick, activeRule.rule.timing);
        }
      }

      this.controlBytes[i] = working.c & ~CONTROL_BYTE_MASKS.RESERVED;
      output[outputIndex] = working.r;
      output[outputIndex + 1] = working.g;
      output[outputIndex + 2] = working.b;
      output[outputIndex + 3] = working.a;
    }
    if (this.onFrameRendered) {
      this.onFrameRendered();
    }
  }
}
